app/src/main/java/com/github/se/eduverse/model/folder/FolderViewModel.kt

```diff
@@ -0,0 +1,95 @@
+package com.github.se.eduverse.model.folder
+
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+class FolderViewModel(val repository: FolderRepository) : ViewModel() {
+
+  private val _existingFolders: MutableStateFlow<MutableList<Folder>> =
+      MutableStateFlow(
+          repository
+              .getFolders(
+                  {},
+                  { Log.e("FolderViewModel", "Exception $it while trying to load the folders") })
+              .toMutableList())
+  val existingFolders: StateFlow<MutableList<Folder>> = _existingFolders
+
+  var activeFolder: MutableStateFlow<Folder?> = MutableStateFlow(null)
+
+  /**
+   * Sort the array of files of the active folder.
+   *
+   * @param filter the filter to apply, as defined in enum FilterTypes in model/folder/Folder.kt
+   */
+  fun sortBy(filter: FilterTypes) {
+    when (filter) {
+      FilterTypes.NAME -> activeFolder.value?.files?.sortBy { it.name }
+      FilterTypes.CREATION_UP -> activeFolder.value?.files?.sortBy { it.creationTime.timeInMillis }
+      FilterTypes.CREATION_DOWN ->
+          activeFolder.value?.files?.sortBy { -it.creationTime.timeInMillis }
+      FilterTypes.ACCESS_RECENT -> activeFolder.value?.files?.sortBy { -it.lastAccess.timeInMillis }
+      FilterTypes.ACCESS_OLD -> activeFolder.value?.files?.sortBy { it.lastAccess.timeInMillis }
+      FilterTypes.ACCESS_MOST -> activeFolder.value?.files?.sortBy { -it.numberAccess }
+      FilterTypes.ACCESS_LEAST -> activeFolder.value?.files?.sortBy { it.numberAccess }
+      else -> throw NotImplementedError("The sort method is not up-to-date")
+    }
+  }
+
+  /**
+   * Add a folder to the list of existing folders.
+   *
+   * @param folder the folder to add
+   */
+  fun addFolder(folder: Folder) {
+    _existingFolders.value.add(folder)
+    repository.addFolder(
+        folder,
+        {},
+        {
+          Log.e(
+              "FolderViewModel",
+              "Exception $it while trying to add folder ${folder.name} to the repository")
+        })
+  }
+
+  /**
+   * Remove a folder from the list of existing folders.
+   *
+   * @param folder the folder to remove
+   */
+  fun deleteFolder(folder: Folder) {
```

Comment:
Same as for add folder the logic of the variables updates should be defined inside the on success lambda